realm/realm-library/src/androidTest/java/io/realm/SortTest.java

```diff
@@ -559,18 +564,70 @@ public void sortByLongDistinctByInt() {
         // (2, 1, "B")
         // (1, 1, "A)
         RealmResults<AnnotationIndexTypes> results1 = realm.where(AnnotationIndexTypes.class)
-                .findAllSorted(AnnotationIndexTypes.FIELD_INDEX_LONG, Sort.DESCENDING);
+                .sort(AnnotationIndexTypes.FIELD_INDEX_LONG, Sort.DESCENDING)
+                .findAll();
         assertEquals(3, results1.size());
         assertEquals(3, results1.get(0).getIndexLong());
 
         // After distinct:
         // (3, 1, "C")
-        RealmResults<AnnotationIndexTypes> results2 =  results1.where().distinct(AnnotationIndexTypes.FIELD_INDEX_INT);
+        RealmResults<AnnotationIndexTypes> results2 =  results1.where().distinctValues(AnnotationIndexTypes.FIELD_INDEX_INT).findAll();
         assertEquals(1, results2.size());
         assertEquals("C", results2.get(0).getIndexString());
         assertEquals(3, results2.get(0).getIndexLong());
     }
 
+    @Test
+    public void sortAndDistinctMixed() {
+        // Dataset:
+        // (FIELD_INDEX_LONG, FIELD_INDEX_INT, FIELD_INDEX_STRING)
+        // (1, 1, "A")
+        // (2, 1, "B")
+        // (3, 1, "C")
+        // Depending on the sorting, distinct should pick the first element encountered.
+        // The order of sort/distinct in the query should not matter
+
+        // Case 1: Selecting highest numbers
+        RealmResults<AnnotationIndexTypes> results1a = realm.where(AnnotationIndexTypes.class)
+                .sort(AnnotationIndexTypes.FIELD_INDEX_LONG, Sort.DESCENDING)
+                .distinctValues(AnnotationIndexTypes.FIELD_INDEX_INT)
+                .findAll();
+        assertEquals(1, results1a.size());
+        assertEquals(3, results1a.get(0).getIndexLong());
+
+        RealmResults<AnnotationIndexTypes> results1b = realm.where(AnnotationIndexTypes.class)
```

Comment:
This seems to be against my instinct. So the sorting is still evaluated first then distinct?
Another question is how does `or()` and `beginGroup()`/`endGroup()` work with `sort()` and `distinct()`?

Reply:
Sort/distinct just saves the descriptor and applies it when you create the `OsResults` object, so the order doesn't matter for that reason.

But semantically I would say that doing distinct before sorting would result in "undefined" in terms of which objects are returned in case of similar value. With sorting first you pick the first encountered after sorting. 

So in this case, "undefined" just means do whatever would have happened if you did sort first. That looks acceptable to me?

What is slightly more annoying is that you can put sort/distinct anywhere, even where it doesn't make a lot of sense:

```
query.equalTo("field", 1).or().distinct("field").equalTo("field", 2).findAll()`
query.equalTo("field", 1).or().beginGroup().sort("field").equalTo("field", 2).endGroup().findAll()`
```

I don't think we can do much about this currently as it would require a more extensive refactoring of the entire query system. On the other side, as long as we don't support subqueries then it isn't directly harmful either from what I can tell?

Reply:
what about force `sort()`/`distinct()` to be called at the end?
```
query.equalTo().sort().findAll(); // good
query.sort().equalTo() // throws
query.sort().distinct() // good
```

Reply:
Yes, that was my idea as well, but without any help from the type system, I think that would actually be worse than allowing it anywhere? Especially since it only impacts readability if you have it elsewhere?

Reply:
OK ... just document this in the `sort()`/`distinctValues()` then. We can make a proper design in the next major release.